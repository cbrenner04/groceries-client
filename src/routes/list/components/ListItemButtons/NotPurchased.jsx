import React from 'react';
import PropTypes from 'prop-types';
import { ButtonGroup } from 'react-bootstrap';

import { listTypeToSnakeCase } from '../../../../utils/format';
import { Bookmark, Complete, Edit, Trash } from '../../../../components/ActionButtons';

function NotPurchasedItemButtons(props) {
  const listItemPath = () => {
    const listId = props.item[`${listTypeToSnakeCase(props.listType)}_id`];
    return `/lists/${listId}/${listTypeToSnakeCase(props.listType)}_items`;
  };

  const handleRead = () => props.handleReadOfItem(props.item);
  const handleUnRead = () => props.handleUnReadOfItem(props.item);

  return (
    <ButtonGroup className="float-right">
      {props.listType === 'BookList' && (
        <Bookmark handleClick={props.item.read ? handleUnRead : handleRead} read={props.item.read} />
      )}
      <Complete handleClick={() => props.handlePurchaseOfItem(props.item)} />
      <Edit to={`${listItemPath()}/${props.item.id}/edit`} />
      <Trash handleClick={() => props.handleItemDelete(props.item)} />
    </ButtonGroup>
  );
}

NotPurchasedItemButtons.propTypes = {
  listType: PropTypes.string.isRequired,
  item: PropTypes.shape({
    id: PropTypes.number.isRequired,
    read: PropTypes.bool,
  }).isRequired,
  handlePurchaseOfItem: PropTypes.func.isRequired,
  handleItemDelete: PropTypes.func.isRequired,
  handleReadOfItem: PropTypes.func.isRequired,
  handleUnReadOfItem: PropTypes.func.isRequired,
};

export default NotPurchasedItemButtons;
