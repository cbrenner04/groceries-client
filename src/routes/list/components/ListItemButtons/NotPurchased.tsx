import React from 'react';
import PropTypes from 'prop-types';
import { ButtonGroup } from 'react-bootstrap';

import { Bookmark, Complete, EditButton, Trash } from '../../../../components/ActionButtons';
import { listItem } from '../../../../prop-types';

function NotPurchasedItemButtons(props) {
  return (
    <ButtonGroup className={`${props.multiSelect ? 'list-item-buttons' : ''} float-end`}>
      {props.listType === 'BookList' && (
        <Bookmark
          handleClick={() => props.toggleItemRead(props.item)}
          read={props.item.read}
          testID={`not-purchased-item-${props.item.read ? 'unread' : 'read'}-${props.item.id}`}
          disabled={props.pending}
        />
      )}
      <Complete
        handleClick={() => props.handlePurchaseOfItem(props.item)}
        testID={`not-purchased-item-complete-${props.item.id}`}
        disabled={props.pending}
      />
      <EditButton
        handleClick={() => props.handleItemEdit(props.item)}
        testID={`not-purchased-item-edit-${props.item.id}`}
        disabled={props.pending}
      />
      <Trash
        handleClick={() => props.handleItemDelete(props.item)}
        testID={`not-purchased-item-delete-${props.item.id}`}
        disabled={props.pending}
      />
    </ButtonGroup>
  );
}

NotPurchasedItemButtons.propTypes = {
  listType: PropTypes.string.isRequired,
  item: listItem.isRequired,
  handlePurchaseOfItem: PropTypes.func.isRequired,
  handleItemDelete: PropTypes.func.isRequired,
  toggleItemRead: PropTypes.func.isRequired,
  handleItemEdit: PropTypes.func.isRequired,
  pending: PropTypes.bool.isRequired,
  multiSelect: PropTypes.bool,
};

export default NotPurchasedItemButtons;
