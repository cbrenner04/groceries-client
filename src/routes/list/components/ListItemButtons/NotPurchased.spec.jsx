import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';

import NotPurchased from './NotPurchased';

describe('NotPurchased', () => {
  const history = createMemoryHistory();
  let props;

  const renderNotPurchased = (localProps) => {
    return render(
      <Router history={history}>
        <NotPurchased {...localProps} />
      </Router>,
    );
  };

  beforeEach(() => {
    props = {
      item: {
        grocery_list_id: 1,
        id: 1,
        read: true,
      },
      purchased: false,
      handleItemDelete: jest.fn(),
      handlePurchaseOfItem: jest.fn(),
      handleReadOfItem: jest.fn(),
      handleUnReadOfItem: jest.fn(),
      handleItemUnPurchase: jest.fn(),
      listType: 'GroceryList',
    };
  });

  it('renders Bookmark when listType is BookList', () => {
    props.listType = 'BookList';
    props.item.book_list_id = 1;
    const { container, getAllByRole } = renderNotPurchased(props);

    expect(container).toMatchSnapshot();
    expect(getAllByRole('button')[0].firstChild).toHaveClass('fa-bookmark');
  });

  it('does not render Bookmark when listType is not BookList', () => {
    props.listType = 'GroceryList';
    const { container, getAllByRole } = renderNotPurchased(props);

    expect(container).toMatchSnapshot();
    getAllByRole('button').forEach((button) => {
      expect(button).not.toHaveClass('fa-bookmark');
    });
  });

  it('calls handleReadOfItem when listType is BookList and read is false and Bookmark is clicked', () => {
    props.listType = 'BookList';
    props.item.read = false;
    const { getAllByRole } = renderNotPurchased(props);

    fireEvent.click(getAllByRole('button')[0]);

    expect(props.handleReadOfItem).toHaveBeenCalledWith(props.item);
  });

  it('calls handleUnReadOfItem when listType is BookList and read is true and Bookmark is clicked', () => {
    props.listType = 'BookList';
    props.item.read = true;
    const { getAllByRole } = renderNotPurchased(props);

    fireEvent.click(getAllByRole('button')[0]);

    expect(props.handleUnReadOfItem).toHaveBeenCalledWith(props.item);
  });

  it('calls handlePurchaseOfItem when Complete is clicked', () => {
    const { getAllByRole } = renderNotPurchased(props);

    fireEvent.click(getAllByRole('button')[0]);

    expect(props.handlePurchaseOfItem).toHaveBeenCalledWith(props.item);
  });

  it('navigates to edit list item when Edit is clicked', () => {
    const { getByRole } = renderNotPurchased(props);

    fireEvent.click(getByRole('link'));

    expect(history.location.pathname).toBe('/lists/1/grocery_list_items/1/edit');
  });

  it('calls handleItemDelete when Trash is clicked', () => {
    const { getAllByRole } = renderNotPurchased(props);

    fireEvent.click(getAllByRole('button')[1]);

    expect(props.handleItemDelete).toHaveBeenCalledWith(props.item);
  });
});
