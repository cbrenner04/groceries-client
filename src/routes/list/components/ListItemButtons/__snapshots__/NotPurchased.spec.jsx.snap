// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotPurchased does not render Bookmark when listType is not BookList 1`] = `
<div>
  <div
    class="float-right btn-group"
    role="group"
  >
    <button
      class="p-0 mr-3  btn btn-link"
      data-test-id="not-purchased-item-complete-id1"
      type="button"
    >
      <i
        class="fa fa-check fa-2x text-success"
      />
    </button>
    <button
      class="p-0 mr-3 btn btn-link"
      data-test-id="not-purchased-item-edit-id1"
      type="button"
    >
      <i
        class="fa fa-edit fa-2x text-warning"
      />
    </button>
    <button
      class="p-0 btn btn-link"
      data-test-id="not-purchased-item-delete-id1"
      type="button"
    >
      <i
        class="fa fa-trash fa-2x text-danger"
      />
    </button>
  </div>
</div>
`;

exports[`NotPurchased renders Bookmark when listType is BookList 1`] = `
<div>
  <div
    class="float-right btn-group"
    role="group"
  >
    <button
      class="p-0 mr-3 btn btn-link"
      data-test-id="not-purchased-item-unread-id1"
      type="button"
    >
      <i
        class="fas fa-bookmark fa-2x text-info"
      />
    </button>
    <button
      class="p-0 mr-3  btn btn-link"
      data-test-id="not-purchased-item-complete-id1"
      type="button"
    >
      <i
        class="fa fa-check fa-2x text-success"
      />
    </button>
    <button
      class="p-0 mr-3 btn btn-link"
      data-test-id="not-purchased-item-edit-id1"
      type="button"
    >
      <i
        class="fa fa-edit fa-2x text-warning"
      />
    </button>
    <button
      class="p-0 btn btn-link"
      data-test-id="not-purchased-item-delete-id1"
      type="button"
    >
      <i
        class="fa fa-trash fa-2x text-danger"
      />
    </button>
  </div>
</div>
`;
